src/main/java/com/github/manolo8/darkbot/extensions/plugins/PluginUpdater.java

```diff
@@ -0,0 +1,288 @@
+        protected Void doInBackground() {
+            updateTasks.forEach(SwingWorker::execute);
+            updateTasks.forEach(UpdateTask::waitUntilDone);
+
+            pluginHandler.updatePlugins();
+
+            updateTasks.forEach(task -> {
+                if (task.failed)
+                   failedUpdates.append(task.plugin.getName()).append(" failed to update\n");
+            });
+            return null;
+        }
+    }
+
+    private class UpdateTask extends SwingWorker<Void, Void> {
+
+        private final StringBuilder failedUpdate = new StringBuilder();
+
+        private final Plugin plugin;
+        private final UpdateAllTask updateAllTask;
+        private final ProgressManager progressManager;
+
+        private boolean failed;
+
+        UpdateTask(Plugin plugin) {
+            this(plugin, null);
+        }
+
+        UpdateTask(Plugin plugin, UpdateAllTask updateAllTask) {
+            this.plugin = plugin;
+            this.updateAllTask = updateAllTask;
+            PluginCard card = pluginDisplay.getPluginCard(plugin);
+            this.progressManager = new ProgressManager(card, this);
+        }
+
+        private void waitUntilDone() {
+            try {
+                get();
+            } catch (InterruptedException | ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        protected void done() {
+            plugin.setUpdateStatus(failed ? Plugin.UpdateStatus.FAILED : Plugin.UpdateStatus.UP_TO_DATE);
+            if (updateAllTask != null || failedUpdate.length() <= 0) return;
+
+            Popups.showMessageAsync("Update failed", failedUpdate, JOptionPane.ERROR_MESSAGE);
+        }
+
+        @Override
+        protected Void doInBackground() throws Exception {
+            progressManager.setUpdateProgress(ProgressManager.State.STARTING);
+
+            if (plugin.getUpdateDefinition() == null)
+                plugin.setUpdateDefinition(findUpdate(plugin.getDefinition()));
+
+            try (InputStream is = plugin.getUpdateDefinition().download.openConnection().getInputStream()) {
+                progressManager.setUpdateProgress(ProgressManager.State.SAVING_OLD);
+
+                FileUtils.createDirectory(PluginHandler.PLUGIN_OLD_PATH);
+                Files.copy(plugin.getFile().toPath(), PluginHandler.PLUGIN_OLD_PATH.resolve(plugin.getFile().getName()), StandardCopyOption.REPLACE_EXISTING);
+                progressManager.setUpdateProgress(ProgressManager.State.DOWNLOADING);
+
+                FileUtils.createDirectory(PluginHandler.PLUGIN_UPDATE_PATH);
+                Files.copy(is, PluginHandler.PLUGIN_UPDATE_PATH.resolve(plugin.getFile().getName()), StandardCopyOption.REPLACE_EXISTING);
+                progressManager.setUpdateProgress(ProgressManager.State.WAITING_RELOAD);
+
+                if (updateAllTask != null) return null;
+
+                progressManager.setUpdateProgress(ProgressManager.State.RELOADING);
```

Comment:
You're setting waiting reload & reloading at the same time, consider just doing just one of the 2